components/cloud/aws/src/s3.rs

```diff
@@ -198,20 +221,63 @@ impl S3Storage {
         Ok(S3Storage { config, client })
     }
 
+    fn maybe_assume_role<P, D>(
+        config: Config,
+        cred_provider: P,
+        dispatcher: D,
+    ) -> io::Result<S3Storage>
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        if let Some(role_arn) = config.role_arn.clone() {
```

Comment:
How about:
```suggestion
        if config.role_arn.is_some() {
```
Then we clone it at...